src/os/bluestore/BlueStore.h

```diff
@@ -490,39 +511,190 @@ class BlueStore : public ObjectStore,
+      if (near) {
+	auto q = buffer_lru.iterator_to(*near);
+	buffer_lru.insert(q, *b);
+      } else if (level > 0) {
+	buffer_lru.push_front(*b);
+      } else {
+	buffer_lru.push_back(*b);
+      }
+      buffer_size += b->length;
+    }
+    int _rm_buffer(Buffer *b) override {
+      buffer_size -= b->length;
+      auto q = buffer_lru.iterator_to(*b);
+      buffer_lru.erase(q);
+      return 1;
+    }
+    void _adjust_buffer_size(Buffer *b, int64_t delta) override {
+      buffer_size += delta;
+    }
+    void _touch_buffer(Buffer *b) override {
+      auto p = buffer_lru.iterator_to(*b);
+      buffer_lru.erase(p);
+      buffer_lru.push_front(*b);
+      _audit("_touch_buffer end");
+    }
+
+    void trim(uint64_t onode_max, uint64_t buffer_max) override;
+
+#ifdef DEBUG_CACHE
+    void _audit(const char *s) override;
+#endif
+  };
+
+  // 2Q cache for buffers, LRU for onodes
+  struct TwoQCache : public Cache {
+  private:
+    // stick with LRU for onodes for now (fixme?)
     typedef boost::intrusive::list<
       Onode,
       boost::intrusive::member_hook<
         Onode,
 	boost::intrusive::list_member_hook<>,
 	&Onode::lru_item> > onode_lru_list_t;
+    typedef boost::intrusive::list<
+      Buffer,
+      boost::intrusive::member_hook<
+	Buffer,
+	boost::intrusive::list_member_hook<>,
+	&Buffer::lru_item> > buffer_list_t;
 
-    std::mutex lock;                ///< protect lru and other structures
-    buffer_lru_list_t buffer_lru;
-    uint64_t buffer_size = 0;
     onode_lru_list_t onode_lru;
 
-    void _touch_onode(OnodeRef& o);
+    buffer_list_t buffer_hot;      //< "Am" hot buffers
+    buffer_list_t buffer_warm_in;  //< "A1in" newly warm buffers
+    buffer_list_t buffer_warm_out; //< "A1out" empty buffers we've evicted
+    uint64_t buffer_bytes = 0;        //< bytes
 
-    void _touch_buffer(Buffer *b) {
-      auto p = buffer_lru.iterator_to(*b);
-      buffer_lru.erase(p);
-      buffer_lru.push_front(*b);
-      _audit_lru("_touch_buffer end");
+    enum {
+      BUFFER_NEW = 0,
+      BUFFER_WARM_IN,   ///< in buffer_warm_in
+      BUFFER_WARM_OUT,  ///< in buffer_warm_out
+      BUFFER_HOT,       ///< in buffer_hot
+    };
+
+  public:
+    void _add_onode(OnodeRef& o, int level) override {
+      if (level > 0)
+	onode_lru.push_front(*o);
+      else
+	onode_lru.push_back(*o);
+    }
+    void _rm_onode(OnodeRef& o) override {
+      auto q = onode_lru.iterator_to(*o);
+      onode_lru.erase(q);
+    }
+    void _touch_onode(OnodeRef& o) override;
+
+    void _add_buffer(Buffer *b, int level, Buffer *near) override;
+
+    int _rm_buffer(Buffer *b) override {
+      if (!b->is_empty()) {
+	buffer_bytes -= b->length;
+      }
+      switch (b->cache_private) {
+      case BUFFER_WARM_IN:
+	buffer_warm_in.erase(buffer_warm_in.iterator_to(*b));
+	return 1;
+      case BUFFER_WARM_OUT:
+	buffer_warm_out.erase(buffer_warm_out.iterator_to(*b));
+	return 2;
```

Comment:
It's pretty unclear what values 1 & 2 means here. From the comments in the header this should be some level used for replacement. In _add_buffer right? But as far as I can see _add_buffer cares of level values that are greater or less than zero, not 1 or 2. Moreover I can't see any assignment for _rm_buffer's return value hence this stuff looks useless at the moment.